### PR TITLE
User approval to resume from user db backup

### PR DESCRIPTION
**Pull request by dimyG**
Addresses issue #5221 

* If there is an ongoing upgrade and the user tries to login he is prompted to resume from the latest user db backup. In that case, the frontend should show two options to the user: 1. Resume from backup  2. Abort login.
* a resume_from_backup argument was added to user login endpoint
* a test case was added

Closes #(issue_number)

## Checklist

- [ ] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

### CODE REVIEW

File: rotkehlchen/db/dbhandler.py [hunk 1]
```diff
@@ -209,6 +214,7 @@ def __init__(
         self.get_or_create_evm_token_lock = Semaphore()
         self.password = password
         self._connect()
+        self.resume_from_backup = resume_from_backup  # before calling _check_unfinished_upgrades
```
Comment (nebolax) [inline rotkehlchen/db/dbhandler.py:217]: What is the reason for storing this as a variable? Isn't it used just a single time?

File: rotkehlchen/rotkehlchen.py [hunk 2]
```diff
@@ -265,10 +266,11 @@ def unlock_user(
             sync_approval=sync_approval,
             sync_database=sync_database,
             initial_settings=initial_settings,
+            resume_from_backup=resume_from_backup,
         )
 
         # unlock or create the DB
-        self.user_directory = self.data.unlock(user, password, create_new, initial_settings)
+        self.user_directory = self.data.unlock(user, password, create_new, initial_settings, resume_from_backup)  # noqa: E501
```
Comment (nebolax) [inline rotkehlchen/rotkehlchen.py:273]: kwargs could be added here and it this call could be split in multiple lines

**Review by nebolax — COMMENTED** [review]
Have you checked whether the same functionality can be achieved with globaldb upgrades?

**Comment by dimyG** [comment]
> Have you checked whether the same functionality can be achieved with globaldb upgrades?

Probably it can. The difference is that the globaldb is initialized at startup and it is outside of the api. There is no api call for that so the approval should come as a backend initialization parameter (similar to sqlite_instructions). Not quite sure though yet, how we can properly catch an exception and return something to the frontend because as I mentioned were are outside of the api.

File: rotkehlchen/errors/api.py [hunk 3]
```diff
@@ -7,7 +7,7 @@ class RotkehlchenPermissionError(Exception):
     The payload contains information to be shown to the user by the frontend so
     they can decide what to do
     """
-    def __init__(self, error_message: str, payload: Optional[dict[str, Any]]) -> None:
+    def __init__(self, error_message: str, payload: Optional[dict[str, Any]] = None) -> None:
```
Comment (LefterisJP) [inline rotkehlchen/errors/api.py:10]: Why do you need to add the default value here? Explicit is better than implicit. What does this have to do with the goal of your PR?

File: rotkehlchen/tests/api/test_users.py [hunk 4]
```diff
@@ -631,6 +631,46 @@ def test_user_login(rotkehlchen_api_server, username, db_password, data_dir):
...
+    data = {'password': db_password, 'sync_approval': 'unknown', 'async_query': True, 'resume_from_backup': False}  # noqa: E501
+    response = requests.post(
+        api_url_for(rotkehlchen_api_server, 'usersbynameresource', name=username),
+        json=data,
+    )
+    task_id = assert_ok_async_response(response)
+    response_data = wait_for_async_task(rotkehlchen_api_server, task_id)
+    assert_error_async_response(
+        response_data=response_data,
+        contained_in_msg='Resume from a backup or quit login',
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_users.py:666]: you will need to edit this after changing the message as I asked above

File: rotkehlchen/tests/api/test_users.py [hunk 5]
```diff
@@ -631,6 +631,46 @@ def test_user_login(rotkehlchen_api_server, username, db_password, data_dir):
...
+        status_code=HTTPStatus.MULTIPLE_CHOICES,
+        result_exists=True,
+    )
+    # Notice: Any login after this point will require a user db backup to be resumed from.
+    # If you need to login without a backup write your test case before the point at which
+    # the user db upgrade status is set to ongoing.
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_users.py:672]: Why add this comment and not just resume from a backup normally? This also should be part ofthe test otherwise the test is incomplete.

File: rotkehlchen/tests/fixtures/rotkehlchen.py [hunk 6]
```diff
@@ -165,11 +166,23 @@ def mock_get_settings(_cursor) -> DBSettings:
 
     original_unlock = rotki.data.unlock
 
-    def augmented_unlock(user, password, create_new, initial_settings):
+    def augmented_unlock(
+            username: str,
+            password: str,
+            create_new: bool,
+            initial_settings: Optional[ModifiableDBSettings] = None,
+            resume_from_backup: bool = True,
```
Comment (LefterisJP) [inline rotkehlchen/tests/fixtures/rotkehlchen.py:174]: same thing as before for this argument

**Review by LefterisJP — COMMENTED** [review]
How is the user prompted in this case? Shouldn't the frontend also make any changes here so that they:

(1) show when there is a problem and we need it to relogin with resume_from_backup = True
(2) Actually call login again with resume_from_backup=Tre

**Comment by dimyG** [comment]
> 

Yes, frontend changes are required. I'm thinking it like this: 

- The user logs in as usual.
- The frontend sends a request to the login endpoint with resume_from_backup=False (or not sending it at all, it defaults to False)
- If everything is ok and no ongoing upgrade is happening, user logs in normally. 
- If there is an ongoing upgrade the login endpoint returns the message "`The encrypted database is in a half-upgraded state. Either resume from a backup or solve the issue manually.`" with an http status code 300 (multiple choices). 
- In that case, the front end should read this and show a message to the user while providing two buttons. One is "Resume from latest backup" and one is "Abort login" (or something like that). 
- If the user presses the "Resume from latest backup" the frontend sends a normal request to the login endpoint but with the resume_from_backup=True. The user logins with the backup user db. 
- If the user presses "Abort login" no request is made to the backend. He is just shown a message to fix manually or something and he is back to the normal login view.

File: rotkehlchen/db/dbhandler.py [hunk 7]
```diff
@@ -233,7 +238,16 @@ def _check_unfinished_upgrades(self) -> None:
         if ongoing_upgrade_from_version is None:
             return  # We are all good
 
-        # Otherwise replace the db with a backup and reconnect
+        # if there is an unfinished upgrade, check user approval to resume from backup
+        if resume_from_backup is False:
+            raise RotkehlchenPermissionError(
+                error_message='The encrypted database is in a half-upgraded state. '
+                              'Either resume from a backup or solve the issue manually.',
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:245]: It's a pity we don't have a way of enforcing this via a linter but this is not how we handle multiline strings. Please do this instead:

Also please notice all the changes my diff had in the last review.

`half -> semi`

```suggestion
                error_message=(
                    'The encrypted database is in a semi upgraded state. '
                    'Either resume from a backup or solve the issue manually.'
                ),
```

or you can also just make it all ine one line if its' not too huge and add `#noqa: E501` at the end

File: rotkehlchen/tests/api/test_users.py [hunk 8]
```diff
@@ -631,6 +634,81 @@ def test_user_login(rotkehlchen_api_server, username, db_password, data_dir):
...
+    response = requests.post(
+        api_url_for(rotkehlchen_api_server, 'usersbynameresource', name=username),
+        json=data,
+    )
+    # And make sure it works
+    task_id = assert_ok_async_response(response)
+    result = wait_for_async_task_with_result(rotkehlchen_api_server, task_id)
+    check_proper_unlock_result(result)
+    assert rotki.user_is_logged_in is True
+    users_data = check_user_status(rotkehlchen_api_server)
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_users.py:699]: here you should also check tha this is the DB with the `VALUES("is_backup", "Yes")` in its settings. You actually went into the trouble of doing it to identify the DB properly so I suppose you just forgot to check it here?

**Review by LefterisJP — COMMENTED** [review]
Okay You will also need to:

1. Edit the api docs to explain the new way login should work and what the new arguments do and what their possible valies are
2. Coordinate with the frontend to let them know what they need to do to complete the issue from their side. Your suggestion [here](https://github.com/rotki/rotki/pull/6131#issuecomment-1559837107) makes sense to me.

File: docs/api.rst [hunk 9]
```diff
@@ -235,7 +237,11 @@ Handling user creation, sign-in, log-out and querying
...
+        - Possibility of syncing exists and the login was sent with sync_approval set to ``"unknown"``. Consumer of api must resend with ``"yes"`` or ``"no"``. In this case the result will contain an object with a payload for the message under the ``result`` key and the message under the ``message`` key. The payload has the following keys: ``local_size``, ``remote_size``, ``local_last_modified``, ``remote_last_modified``.
+
    :statuscode 400: Provided JSON is in some way malformed
    :statuscode 401: Provided password is wrong for the user or some other authentication error.
    :statuscode 409: Another user is already logged in. User does not exist. There was a fatal error during the upgrade of the DB. Permission error while trying to access the directory where rotki saves data.
```
Comment (LefterisJP) [inline docs/api.rst:247]: you will need to add the possible statuscode of 300 here too
